lib/widgets/login.dart

```diff
@@ -373,8 +373,15 @@ class _PasswordLoginPageState extends State<PasswordLoginPage> {
       decoration: InputDecoration(
         labelText: 'Password',
         helperText: kLayoutPinningHelperText,
-        suffixIcon: Semantics(label: 'Hide password', toggled: _obscurePassword,
+        // TODO(material-3): Simplify away `Semantics` by using IconButton's
+        //   M3-only params `isSelected` / `selectedIcon`, after fixing
+        //   https://github.com/flutter/flutter/issues/127145 . (Also, the
+        //   `Semantics` seen here would misbehave in M3 for reasons
+        //   involving a `Semantics` with `container: true` in an underlying
+        //   [ButtonStyleButton].)
```

Comment:
Is there an issue to link to for this?

Reply:
No, I don't think so. The change in behavior with M3 surprised me, but I haven't yet put my finger on an upstream bug. And as you saw, once https://github.com/flutter/flutter/issues/127145 is in, we can get it right in M3 with simpler code anyway. 🙂

Here are some observations about what happens if I set `useMaterial3: true` near our app root:

`IconButton` builds something it calls [`_IconButtonM3`](https://github.com/flutter/flutter/blob/7b67aa587/packages/flutter/lib/src/material/icon_button.dart#L869), and that's a class that extends [`ButtonStyleButton`](https://api.flutter.dev/flutter/material/ButtonStyleButton-class.html). And [`ButtonStyleButton` builds a `Semantics` with `container: true`](https://github.com/flutter/flutter/blob/7b67aa587/packages/flutter/lib/src/material/button_style_button.dart#L427), to introduce a new node in the Semantics tree ([doc](https://api.flutter.dev/flutter/widgets/Semantics/container.html)). I guess that could be a reasonable choice, I'm not sure. 🙂 I can focus the button with VoiceOver, and it just says: "Hide password. Button."—which means my `toggled` value is ignored. I found three separate ways I could fix the issue, to where it's still focusable and it says "Hide password. Switch button. On. Double-tap to toggle setting.". Those ways are:

- If I remove that `container: true` in the `ButtonStyleButton` code. Or,
- If I wrap my `Semantics(toggled: _obscurePassword,` with a [`MergeSemantics`](https://api.flutter.dev/flutter/widgets/MergeSemantics-class.html). Or,
- My example patch in that issue I filed: https://github.com/flutter/flutter/issues/127145#issuecomment-1553767753

If I copy this chunk of code (the one beginning with `Semantics(toggled: _obscurePassword`) to somewhere outside the password text field (say, between it and the "Log in" button), then the behaviors I've described so far replicate there too. I tested that because I noticed something odd happening near the original show/hide button: before applying one of the three fixes I listed above, a _different_ focusable node would get the wrong text: the one for the whole password text field, before I drill down to (traverse forward to?) the show/hide button. VoiceOver would say, "Password. 1. Text field. Double-tap to edit"—or, for the show/hide button's other state, "Password. 0. Text field. Double-tap to edit". In that case, is the framework wanting to present `suffixIcon` as part of the _meaning_ of the whole text field? ([This code](https://github.com/flutter/flutter/blob/7b67aa587/packages/flutter/lib/src/material/input_decorator.dart#L840-L881) looks related…?) But here it's not meant as part of the text field's meaning; it's just a related control we offer.